### ISSUE TITLE
Configuration Option to Override Plugin's Generated Config Name

### ISSUE THREAD

**Issue by Formula350** [BepInEx/BepInEx#225]
**Is your feature request related to a problem? Please describe.**
It's a huuuge pain in the *asterisks* to sort through the config files and try to pair them up to the mod they belong to, when the mod author opt not to change/define a name.... Having to scroll up and down a few times to figure out what the config file's name is.

Here's a friendly face we can use as an example, to not call out a mod author directly... lol
`com.bepis.bepinex.configurationmanager.cfg`
Granted, it's slightly better formatted than some I've come across, but, the fact that the actual mod/plugin file is called `ConfigurationManager.dll`, one intuitively looks for `ConfigurationManager.cfg` 

**Describe the solution you'd like**
Simply an override in the intuitively named `BepInEx.cfg` that will automatically rename the config file by the plugin that's creating it, to whatever the plugin's filename is (respecting whitespace and upper/lowercasing).

Therefore, `ConfigurationManager.dll` will now produce `ConfigurationManager.cfg`.

Something to the tune of this perhaps?
```
## Force each plugin's config file to be named identically to the filename of that specific plugin.
## NOTE: Remember to factor this in while reading tutorials or reporting bugs, to not be confused by
##            any filename differences OR accidentally confuse others while explaining filenames.
# Setting type: Boolean
# Default value: false
OverrideCfgNaming = true
```
*(alternative name: PluginMatchesConfig, NameMatchedFiles, MakeMyLifeEasier, NoMasochisticTendencies, ICanHazSanity `[false = NoThnx; true = YasPl0x]`)*

**Describe alternatives you've considered**
Reaching out to every single mod author to ask them to do whatever it is they need to do (since I don't know, myself), to have their config files generate with the same name as their mod's DLL.

**Additional context**
BONUS ROUND! Maybe also have it automagically plop additional twined-octothorpes *(hashtags, if you rather)* entries on a new line in the config, similar to how I did it in the above example. (this might be a bug, I don't really have a clue! )
Currently some configs of mods I have for Valheim, they display like this *(albeit rare; although, the aforementioned config manager does! lol)*:
```
## Text text text text text.## Moar text text text text text.## Aaaaand text!
# Setting type: stuff
# Default value: some-value
DescriptiveWords = same-or-similar-value
```
Whereas one would expect it'd be formatted....
```
## Text text text text text
## Moar text text text text text
## Aaaaand text!
# Setting type: stuff
# Default value: some-value
DescriptiveWerds = same-or-similar-value
```

**Comment by Windows10CE** [comment]
Hello! The plugins files you're talking about currently use the default config file provided by BepInEx to developers, which uses their plugins GUID (that string starting with `com.` that you see). Plugin developers wouldn't have to do anything on their end if we were to implement an option for this, and the most likely solution on our end would be to give the option of GUID or plugin name (this is not the DLL name, but should be very similar most of the time).

The main problem I see is:
How do we handle migrating between setting values? If a user decides to switch the names after already messing with mod configs, do we just ignore the files that exist and create new ones? Or try to do some kind of migration? Trying to migrate them wouldn't be super obnoxious, but then how do you deal with multiple variations existing?


As for the second problem you describe, I've never seen that personally. Even my own copy of the example you give doesn't seem to display the same behavior, but it's quite possible there's an issue somewhere there.
![image](https://user-images.githubusercontent.com/34900473/113965767-82270180-981d-11eb-87dc-6c5fcfa44cee.png)

You mentioned Valheim in your other issue, so I'll assume that's the main game your playing. On your end I would suggest possibly using a mod manager such as [r2modman](https://valheim.thunderstore.io/package/denikson/BepInExPack_Valheim/) if that's an option, it provides a search for config files and will manage everything for you.

**Comment by ghorsington** [comment]
I'm going to add to what Windows10CE mentioned in relation to file naming.

As he pointed out, the `com.bepis.bepinex.configurationmanager.cfg` name comes from the GUID of the plugin. This is because it is *the only uniquely identifying string for a plugin*. GUIDs are used because they are unique: BepInEx simply will never load two plugins with the same GUID.

The reason for using GUIDs is because all other metadata is not unique. For instance, you can have two plugins, say `Foo.dll` and `Bar.dll` that have the following classes

```cs
// In Foo.dll
[BepInPlugin("org.plugin.foo", "Foo Plugin", "1.0.0")]
public class Plugin : BaseUnityPlugin {}

// In Bar.dll
[BepInPlugin("org.plugin.bar", "BarPlugin", "1.0.0")]
public class Plugin : BaseUnityPlugin {}
```

This means that you cannot name the config file `Plugin.cfg` as the name will now overlap for both plugin classes.

In addition, even DLL filenames are not unique! BepInEx allows you to put each plugin in a subfolder inside `plugins` folder so you can have the following folder setup:

```
plugins
| - Foo
   | - Plugin.dll
| - Bar
    | - Plugin.dll 
```

i.e. there is `Plugin.dll` in each of subfolders.

Depending on mono runtime version shipped with the game, assembly version and internal assembly name, this can be allowed and can be loaded fine by BepInEx. In that case again, naming your config `Plugin.cfg` will cause conflicts.  
Moreover, this approach does not take into account the fact that *one assembly can contain multiple plugins*. Indeed, you can specify as many `BaseUnityPlugin`s as you want in your DLL. This is done to support different variations of the same plugin or to allow to bundle multiple plugins into one single DLL.  
Again, you cannot use DLL name as config name because each assembly can have multiple plugins and the name of each plugin type is not unique.

All in all, GUIDs are always unique. BepInEx will never load plugins with the same GUIDs at the same time.

The best option I see is to allow plugin developers to rename their plugin files. This can be done by making [`ConfigFile.ConfigFilePath`](https://github.com/BepInEx/BepInEx/blob/v5-lts/BepInEx/Configuration/ConfigFile.cs#L58) settable. 
It is then up to the developer to name their config files differently and make sure the name does not collide.

All in all, I shall repeat what Windows10CE mentioned: consider using Configuration Manager or r2modman for config management if you don't feel like editing files manually.

I'll keep the issue open for a little while if you have alternative suggestions on the matter on how to obtain more ""readable"" automatically generated plugin config names that are also unique.

**Comment by Formula350** [comment]
@Windows10CE Thanks to your screenshot, I've determined the "issue" with the linebreaks.... It's because for simple cfg files I don't bother to fire up Notepad++ and let Windows Notepad handle it. However, seems that even though the file is encoded to "Dos/Windows" and is UTF-8, that M$'s Notepad just wants to be... difficult/obnoxious/a troll? lol
Either way, its the lack of **[CR]** between those, as there's only a **[LF]**, where everything else has **[CR][LF]**.
:) So that certainly takes care of *that* side-point! Thanks.

As for using a Mod Manager, I probably would have (and do use FOMM for Fallout series games), but then I saw how painfully simple/easy they were to install on Unity so it seemed pointless. I do use the ConfigManager and that greatly helps, but sometimes I just get that urge to edit the actual file before loading or after exiting (having forgot to modify it before quitting), due to that some settings inherently require a restart for the changes to be reflected -- generally rendering related stuff. *(Unless there's a "vidrestart" console command I can use to force changes??)*



In terms of your points with the actual request...
I thought it ***may*** be easy enough -- *which as a non-coder I can appreciate just how incredibly off-base I can be here* -- for Bep to read the Filename of the plugin its loading and override any internal GUID name setting/field. That way even a user who, lets say downloads a mod that... gives fish arms and legs called... mmm... `Evolution.dll` \*shrug\* Then lets say that for some reason they wanted to go so far as rename that mod to `First.dll`, so that their matching config will be `First.cfg`. They'd be able to do so in this case. The internal naming is ignored regardless.

Migration of configs already generated before the use of this option, to me once more seems straight forward: internally Bep could start writing in an identifier. Frankly, *most* of the plugins I have seem to do this already. A few examples:
```
## Settings file was created by plugin Monster AI Tweaks v0.3.0
## Plugin GUID: aedenthorn.MonsterAITweaks
```
```
## Settings file was created by plugin Hitbox Fix v1.3.0
## Plugin GUID: hitbox.fix
```
I included the last one because this is that person's first mod ever, and it includes the GUID in the config.
So I figure, that would be a sufficient identifier that all mod's would have. 

EXAMPLE:
If my feature is added the Migration process for that "hitbox.fix" mod would be...
**[1]**-Locate file named exactly 'hitbox.fix.cfg` (case sensitive)
**[2]**-Ensure said file contains the line:  `## Plugin GUID: hitbox.fix
**[3]**-Check filename of mob/plugin that the file in Step 1 belongs to; which is `SlopeCombatFix.dll`
**[4]**-Rename file in Step 1 to `SlopeCombatFix.cfg`
**[5**]-*OPTIONAL:* append to the top of the file somewhere: `## Filename Overridden. Original = hitbox.fix.cfg`
DONE - Migration complete!
No additional clutter created from new files existing next to now-obsolete ones!


In comes the points made by @ghorsington
Yes, admittedly, I did not consider the fact that you could theoretically have, in different sub-folders **[a]**two mods that may very well do two different things, yet have the same exact filename. OR **[b]**have the same exact mod, maybe of different versions, in two different folders. OR **[c]**some incredibly crazy coincidence where the internal plugin name (public class) are identical, their functions may otherwise be different, but the user and/or creator just so happened to *also* give it the same filename. 

**[a]** Failover to the config file being named `FileName.FolderName.cfg`, since it sounds like in this situation, the mods would still load because their Plugin Class is different
**[b]** Throw an exception that duplicate plugins were detected and the highest version number is being used. OR if the version string matches, fallback to the newest `Created` date, or should those match, `Modified` date of the file. **IF** all of those match... Either throw a second exception which cancels the process and exit, while displaying problem. Perhaps something to the effect of:
"Configuration Filename Override Error: 100% duplicate plugins found in different directories: <directory1\modname.dll> and <directory2\modname.dll>. Please remove one before launching again."
**[c]** Either fallback to Option A's solution of `FileName.FolderName.cfg` and continue anyways (since I imagine you're saying this is only a problem when trying to do my request during matching "Plugin Class"?), OR cancel and exit, explaining the problem. as with Option B.

As far as I understand it, in your example, the only issue being when it would come down to the filenames, even if internally everything matched. 
However, if internally there are some fields that match between two or more plugins, it sounds like that's not exclusive to my feature request and that there must already be a workaround. 

Even in the case of there being multiple packages inside a single DLL, it still sounds like there would be an internal name to differentiate each config. All we'd need to do is append the Plugin's Filename to the beginning of the Config's Filename.
So that instead of `com.something.person.ModName.cfg` we'd have `Filename.com.something.person.ModName.cfg`.
To me, that's a lot easier to contend with, as there may be a dozen mods contained in a single plugin but now we can visually reference and tie them all to the same file.

.

The long and short of my request basically boils down to that there are situations where the mod maker will not have the knowledge to adhere to proper coding convention, or may not even consider whether or not what they've entered makes it easy for others to pick out the configs. :) 

If this really is way too much hassle than it's worth, I suppose I could always do what just came to me (although it'd be continuous legwork I'd have to do every mod), whereby I have a folder in Bepinex called "configs" (plural), and then I make Shortcuts to every config file, and manually rename *those* to be organized in more me-friendly system.

Thanks for hearing me out, sorry for rambling! heh

**Comment by ghorsington** [comment]
> Migration of configs already generated before the use of this option, to me once more seems straight forward: internally Bep could start writing in an identifier.

In-file identifies are not reliable since they are not mandated to be in there. The reason they were added in the first place was to the human-readable metadata for end-users, not tools. Moreover, with BepInEx 6 being in planning to allow plugins to swap the config file format (i.e. they can write all their configs as JSON if a plugin developer so chooses), the identifier will not be consistent.

> I did not consider the fact that you could theoretically have, in different sub-folders

The issue is less about just plugin placement but the fact that all plugin metadata (type name, assembly name, folder placement) *except the GUID* is not designed to be immutable. There are reasons for why each of them can be changed between plugin versions or between end-users. All of this can manifest in the examples I outlined above, but it is not an exhaustive list.

No matter how you name the plugin, if you add possibly mutable metadata to the name, you'd have to be able to handle migrations like @Windows10CE mentioned. In that case we end up back to issues I outlined earlier.  
All this still does not take into account all the possible processing that has to be gone into it: if you were to change plugin/type name such that the "readable" config name changed from, say `Bar.cfg` to `Foo.cfg`, BepInEx has to scan *all* the configuration files to find the old config to migrate settings from. Again, you can't rely on just identifiers inside the config file, they can be removed and I personally don't want to invalidate the whole configuration file if such identifier gets removed.

All in all, it seems to be that allowing developers to explicitly rename the default plugin configuration file in code seems best here. In that case the name is hardcoded into the plugin and the developer has to handle name migrations if there is a need.

Whatever the solution, I'm moving this discussion from an issue ticket to a Discussions thread.
